File: internal/httpapi/secrets.go
package httpapi

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/go-freya/freya/services/warden/internal/secrets"
)

// RegisterSecrets mounts the secret routes (contracts §secrets).
func (s *Server) RegisterSecrets(d StoryDeps) {
	s.MustHandle("GET", "/api/warden/v1/secrets", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		q := r.URL.Query()
		var folder *string
		if f := q.Get("folder_id"); f != "" {
			folder = &f
		}
		limit, _ := strconv.Atoi(q.Get("limit"))
		page, err := d.Secrets.List(r.Context(), subj, folder, q.Get("cursor"), limit)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, page)
	})
	s.MustHandle("POST", "/api/warden/v1/secrets", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			FolderID    *string         `json:"folder_id"`
			Name        string          `json:"name"`
			Username    string          `json:"username"`
			HostURL     string          `json:"host_url"`
			Description string          `json:"description"`
			Metadata    json.RawMessage `json:"metadata"`
			Password    string          `json:"password"`
			TOTP        string          `json:"totp"`
		}
		if err := DecodeJSON(r, &in, 0); err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Secrets.Create(r.Context(), subj, secrets.Input{FolderID: in.FolderID, Name: in.Name, Username: in.Username, HostURL: in.HostURL, Description: in.Description, Metadata: in.Metadata, Password: in.Password, TOTP: in.TOTP})
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusCreated, out)
	})
	s.MustHandle("GET", "/api/warden/v1/secrets/search", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		q := r.URL.Query()
		limit, _ := strconv.Atoi(q.Get("limit"))
		page, err := d.Secrets.Search(r.Context(), subj, q.Get("q"), q.Get("cursor"), limit)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, page)
	})
	s.MustHandle("GET", "/api/warden/v1/secrets/{id}", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Secrets.Get(r.Context(), subj, r.PathValue("id"))
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("PUT", "/api/warden/v1/secrets/{id}", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			Name        *string         `json:"name"`
			Username    *string         `json:"username"`
			HostURL     *string         `json:"host_url"`
			Description *string         `json:"description"`
			Metadata    json.RawMessage `json:"metadata"`
		}
		if err := DecodeJSON(r, &in, 0); err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Secrets.Update(r.Context(), subj, r.PathValue("id"), secrets.Patch{Name: in.Name, Username: in.Username, HostURL: in.HostURL, Description: in.Description, Metadata: in.Metadata})
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("POST", "/api/warden/v1/secrets/{id}/remove", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		if err := d.Secrets.Delete(r.Context(), subj, r.PathValue("id")); err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		w.WriteHeader(http.StatusNoContent)
	})
	s.MustHandle("POST", "/api/warden/v1/secrets/{id}/move", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			FolderID *string `json:"folder_id"`
		}
		if err := DecodeJSON(r, &in, 0); err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Secrets.Move(r.Context(), subj, r.PathValue("id"), in.FolderID)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("GET", "/api/warden/v1/secrets/{id}/password", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		version, _ := strconv.Atoi(r.URL.Query().Get("version"))
		out, err := d.Secrets.Reveal(r.Context(), subj, r.PathValue("id"), version)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("PUT", "/api/warden/v1/secrets/{id}/password", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			Password string `json:"password"`
			Comment  string `json:"comment"`
		}
		if err := DecodeJSON(r, &in, 0); err != nil {
			Fail(w, r, nil, err)
			return
		}
		v, err := d.Secrets.UpdatePassword(r.Context(), subj, r.PathValue("id"), in.Password, in.Comment)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, map[string]any{"version": v})
	})
	s.MustHandle("GET", "/api/warden/v1/secrets/{id}/versions", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Secrets.Versions(r.Context(), subj, r.PathValue("id"))
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, map[string]any{"items": out})
	})
	s.MustHandle("POST", "/api/warden/v1/secrets/{id}/versions/{version}/restore", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			Comment string `json:"comment"`
		}
		if r.ContentLength != 0 {
			if err := DecodeJSON(r, &in, 0); err != nil {
				Fail(w, r, nil, err)
				return
			}
		}
		version, _ := strconv.Atoi(r.PathValue("version"))
		v, err := d.Secrets.Restore(r.Context(), subj, r.PathValue("id"), version, in.Comment)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, map[string]any{"version": v})
	})
	s.MustHandle("GET", "/api/warden/v1/secrets/{id}/totp", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Secrets.TOTPCode(r.Context(), subj, r.PathValue("id"))
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("PUT", "/api/warden/v1/secrets/{id}/totp", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			TOTP string `json:"totp"`
		}
		if err := DecodeJSON(r, &in, 0); err != nil {
			Fail(w, r, nil, err)
			return
		}
		if err := d.Secrets.SetTOTP(r.Context(), subj, r.PathValue("id"), in.TOTP); err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		w.WriteHeader(http.StatusNoContent)
	})
	s.MustHandle("DELETE", "/api/warden/v1/secrets/{id}/totp", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		if err := d.Secrets.RemoveTOTP(r.Context(), subj, r.PathValue("id")); err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		w.WriteHeader(http.StatusNoContent)
	})
}

File: ui/src/stores/secrets.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '@/api/client'
import type { Page, Secret, SecretInput, SecretVersion } from '@/api/types'

export const useSecrets = defineStore('warden-secrets', () => {
  const items = ref<Secret[]>([])
  const next = ref<string | undefined>()
  const loading = ref(false)
  const error = ref('')
  const folderId = ref<string | null>(null)
  const query = ref('')

  async function list(folder: string | null, cursor?: string): Promise<void> {
    loading.value = true
    error.value = ''
    try {
      const page = await api<Page<Secret>>('GET', 'secrets', undefined, { query: { folder_id: folder ?? undefined, root: folder === null ? true : undefined, cursor, limit: 50 } })
      items.value = cursor ? [...items.value, ...page.items] : page.items
      next.value = page.next
      folderId.value = folder
      query.value = ''
    } catch (e) {
      error.value = (e as Error).message
    } finally {
      loading.value = false
    }
  }

  async function search(q: string, cursor?: string): Promise<void> {
    if (!q.trim()) return list(folderId.value)
    loading.value = true
    error.value = ''
    try {
      const page = await api<Page<Secret>>('GET', 'secrets/search', undefined, { query: { q: q.trim(), cursor, limit: 50 } })
      items.value = cursor ? [...items.value, ...page.items] : page.items
      next.value = page.next
      query.value = q
    } catch (e) {
      error.value = (e as Error).message
    } finally {
      loading.value = false
    }
  }

  async function more(): Promise<void> {
    if (!next.value) return
    if (query.value) return search(query.value, next.value)
    return list(folderId.value, next.value)
  }

  async function refresh(): Promise<void> {
    if (query.value) return search(query.value)
    return list(folderId.value)
  }

  async function get(id: string): Promise<Secret> {
    return api<Secret>('GET', 'secrets/' + id)
  }

  async function create(input: SecretInput): Promise<Secret> {
    const s = await api<Secret>('POST', 'secrets', input)
    await refresh()
    return s
  }

  async function update(id: string, patch: Partial<SecretInput>): Promise<Secret> {
    const s = await api<Secret>('PUT', 'secrets/' + id, patch)
    items.value = items.value.map((i) => (i.id === id ? s : i))
    return s
  }

  async function reveal(id: string, version?: number): Promise<{ password: string; version: number }> {
    return api('GET', 'secrets/' + id + '/password', undefined, { query: { version } })
  }

  async function changePassword(id: string, password: string, comment: string): Promise<number> {
    const out = await api<{ version: number }>('PUT', 'secrets/' + id + '/password', { password, comment })
    await refresh()
    return out.version
  }

  async function versions(id: string): Promise<SecretVersion[]> {
    const out = await api<{ items: SecretVersion[] }>('GET', 'secrets/' + id + '/versions')
    return out.items
  }

  async function restore(id: string, version: number, comment: string): Promise<number> {
    const out = await api<{ version: number }>('POST', 'secrets/' + id + '/versions/' + version + '/restore', { comment })
    await refresh()
    return out.version
  }

  async function move(id: string, folder: string | null): Promise<Secret> {
    const s = await api<Secret>('POST', 'secrets/' + id + '/move', { folder_id: folder })
    await refresh()
    return s
  }

  async function remove(id: string): Promise<void> {
    await api('POST', 'secrets/' + id + '/remove')
    items.value = items.value.filter((i) => i.id !== id)
  }

  async function totp(id: string): Promise<{ code: string; period: number; expires_in: number }> {
    return api('GET', 'secrets/' + id + '/totp')
  }

  async function setTotp(id: string, seed: string): Promise<void> {
    await api('PUT', 'secrets/' + id + '/totp', { totp: seed })
    await refresh()
  }

  async function removeTotp(id: string): Promise<void> {
    await api('DELETE', 'secrets/' + id + '/totp')
    await refresh()
  }

  return { items, next, loading, error, folderId, query, list, search, more, refresh, get, create, update, reveal, changePassword, versions, restore, move, remove, totp, setTotp, removeTotp }
})
